### PR TITLE
Add DeriveKeyPair and apply other minor changes

### PR DESCRIPTION
**Pull request by chris-wood**
This change adds the DeriveKeyPair KEM interface in lieu of the GenerateKeyPair interface. It also updates the test vectors to store only DeriveKeyPair IKM seeds instead of serialized public and private key shares. (Should we keep the private and public keys in addition to the seeds?)

This change also adds the test vectors as an artifact in the repository.

cc @dmcardle, @rozbb, @raphaelrobert, @kjacobs-moz, @blipp to check against the new test vectors. I *think* I covered all the diffs from HPKE, but please shout if I missed something!

### CODE REVIEW

File: crypto.go [hunk 1]
```diff
@@ -237,25 +251,50 @@ func (s ecdhScheme) ID() KEMID {
...
+		return 0x01
+	}
+	panic(fmt.Sprintf("Unsupported curve: %s", s.curve.Params().Name))
+}
+
+func (s ecdhScheme) DeriveKeyPair(ikm []byte) (KEMPrivateKey, KEMPublicKey, error) {
+	idBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(idBuffer, uint16(s.ID()))
+
+	dkp_prk := s.KDF.LabeledExtract(nil, string(idBuffer)+"dkp_prk", ikm)
```
Comment (dmcardle) [inline crypto.go:268]: The `nil` salt translates to `make([]byte, s.hash.Size())` over in ecdhScheme's definition, but the spec says the salt should be `zero(0)`. Is this a discrepancy?

Comment (chris-wood) [reply crypto.go:268]: Nice eye! This isn't a problem, as that behavior part of the HKDF spec (https://tools.ietf.org/html/rfc5869, 2.2).

File: crypto.go [hunk 2]
```diff
@@ -324,56 +372,64 @@ type x25519PublicKey struct {
...
 	return DHKEM_X25519
 }
 
-func (s x25519Scheme) GenerateKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey, error) {
-	priv := &x25519PrivateKey{}
-	_, err := rand.Read(priv.val[:])
+func (s x25519Scheme) DeriveKeyPair(ikm []byte) (KEMPrivateKey, KEMPublicKey, error) {
+	idBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(idBuffer, uint16(s.ID()))
+	dkp_prk := s.KDF.LabeledExtract(nil, string(idBuffer)+"dkp_prk", ikm)
```
Comment (dmcardle) [inline crypto.go:390]: Another question about DeriveKeyPair: is `I2OSP(idBuffer) + "dpk_prk"` the label you want to use? The spec says `I2OSP(kem_id, 2)`.

Comment (chris-wood) [reply crypto.go:390]: I'm going off the changes in this to-be-landed PR: https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/122

Comment (dmcardle) [reply crypto.go:390]: Aha, thanks for the link.

**Comment by dmcardle** [comment]
Success! I can verify that the test vectors pass on my end — that is, mode = base, kem = DHKEM(X25519, SHA256), and kdf = HKDF_SHA256.

**Comment by chris-wood** [comment]
> Success! I can verify that the test vectors pass on my end — that is, mode = base, kem = DHKEM(X25519, SHA256), and kdf = HKDF_SHA256.

Sweet! Have you implemented support for any of the NIST curves?

**Comment by dmcardle** [comment]
> > Success! I can verify that the test vectors pass on my end — that is, mode = base, kem = DHKEM(X25519, SHA256), and kdf = HKDF_SHA256.
> 
> Sweet! Have you implemented support for any of the NIST curves?

No, I think we're only targeting X25519 until we have a need for another curve.

**Comment by chris-wood** [comment]
> No, I think we're only targeting X25519 until we have a need for another curve.

Makes sense. Thanks for clarifying!

**Comment by rozbb** [comment]
Waiting on the dust to settle on the PR before I get to implementing things

File: crypto.go [hunk 3]
```diff
@@ -74,12 +79,17 @@ func (s dhkemScheme) getEphemeralKeyPair(rand io.Reader) (KEMPrivateKey, KEMPubl
...
+
+	return s.group.DeriveKeyPair(ikm)
 }
 
 func (s dhkemScheme) extractAndExpand(dh []byte, kemContext []byte, Nzz int) []byte {
-	prk := s.KDF.LabeledExtract(nil, "dh", dh)
-	return s.KDF.LabeledExpand(prk, "prk", kemContext, Nzz)
+	idBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(idBuffer, uint16(s.ID()))
+	prk := s.group.internalKDF().LabeledExtract(nil, string(idBuffer)+"prk", dh)
```
Comment (dmcardle) [inline crypto.go:91]: Shouldn't the extract label be `string(idBuffer)+"eae_prk"`, based on https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/122?

Comment (chris-wood) [reply crypto.go:91]: Yeah, that PR just changed this morning. I'll update this to match soon. :)

Comment (dmcardle) [reply crypto.go:91]: Aha! No rush. I was just giving my code a second pass vs the spec. I'm still learning the GitHub UI, apparently 🙃

Comment (chris-wood) [reply crypto.go:91]: Hehe, you and me both.

Comment (chris-wood) [reply crypto.go]: @dmcardle updated!

**Comment by chris-wood** [comment]
@rozbb the spec PR landed. Can you please give these test vectors a shot?

**Comment by dmcardle** [comment]
OK, latest test vectors are passing for me!

File: format_vectors.py [hunk 4]
```diff
@@ -3,10 +3,12 @@
 import textwrap
 
 ordered_keys = [
-    "mode", "kemID", "kdfID", "aeadID", "info", "skRm",
-    "skSm", "skEm", "psk", "pskID", "pkRm", "pkSm", "pkEm",
-    "enc", "zz", "key_schedule_context", "secret", "key", "nonce",
-    "exporterSecret",
+    # Mode and ciphersuite parameters
+    "mode", "kemID", "kdfID", "aeadID", "info", "seedR",
```
Comment (rozbb) [inline format_vectors.py:7]: seedR doesn't belong on this line

**Comment by rozbb** [comment]
In the process of comparing and contrasting. One thing that I think needs to be re-added is serialized pubkeys/privkeys. It's way easier to check my DeriveKeyPair impl if I have actual keys to compare to. Also I think this diff wiped out all uses of serialize/deserialize, which is bad from a testing perspective.

**Comment by rozbb** [comment]
Oh, I take that back. It looks like seedE is explicitly compared to a pubkey, i.e., the encapped key. Still, secret key deserialization would be good too.

File: crypto.go [hunk 5]
```diff
@@ -74,12 +79,17 @@ func (s dhkemScheme) getEphemeralKeyPair(rand io.Reader) (KEMPrivateKey, KEMPubl
...
+	return s.group.DeriveKeyPair(ikm)
 }
 
 func (s dhkemScheme) extractAndExpand(dh []byte, kemContext []byte, Nzz int) []byte {
-	prk := s.KDF.LabeledExtract(nil, "dh", dh)
-	return s.KDF.LabeledExpand(prk, "prk", kemContext, Nzz)
+	idBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(idBuffer, uint16(s.ID()))
+	eae_prk := s.group.internalKDF().LabeledExtract(nil, string(idBuffer)+"eae_prk", dh)
+	return s.group.internalKDF().LabeledExpand(eae_prk, "zz", kemContext, Nzz)
```
Comment (rozbb) [inline crypto.go:92]: The spec has KEM_ID || "zz" as the label for LabeledExpand. I think that's an error in the spec, though. No other LabeledExpand calls have KEM binding. When KEM binding is necessary, it's done in LabeledExtract.

**Comment by rozbb** [comment]
Tests pass on all X25519 and P256 test vectors. Aside from my previous comments, this is good to go.

File: format_vectors.py (hunk 4 above)
Comment (chris-wood) [reply format_vectors.py:7]: Oops. Good catch!

**Comment by chris-wood** [comment]
@rozbb I brought back the serialize key material. :)

**Comment by chris-wood** [comment]
Updated to take [#127](https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/127) into account. @bifurcation, please have a look.

**Comment by chris-wood** [comment]
@dmcardle and @rozbb, can you please give these new vectors a shot with your implementation?

File: crypto.go [hunk 6]
```diff
@@ -718,6 +813,12 @@ func (s hkdfScheme) ID() KDFID {
 	panic(fmt.Sprintf("Unsupported hash: %d", s.hash))
 }
 
+func (s hkdfScheme) kemID() []byte {
+	idBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(idBuffer, uint16(s.ID()))
```
Comment (rozbb) [inline crypto.go:818]: This method doesn't make sense. `kemID` shouldn't be part of `hkdfScheme`. I believe this outputs the KDF ID, which would explain why DeriveKeyPair is failing in my KAT

Comment (chris-wood) [reply crypto.go]: Hah, yeah, good point. The intent of putting it on the KDF interface was to say, "give me the suite ID when I'm being used as part of a KEM," but as you point out that's somewhat confusing (and it was using the wrong ID)! I moved things around to address this. Now, each time the implicit suite_id is required, the calling code passes the KEM ID to a helper function to compute this.

**Comment by rozbb** [comment]
Can you update the test vectors to reflect the kemID change?

**Comment by chris-wood** [comment]
> Can you update the test vectors to reflect the kemID change?

I forgot to push that. Stand by!

**Comment by dmcardle** [comment]
Cool, test vectors are passing for me.

File: hpke.go [hunk 7]
```diff
@@ -147,39 +150,40 @@ type contextParameters struct {
...
 		return contextParameters{}, err
 	}
 
-	pskIDHash := suite.KDF.LabeledExtract(nil, "pskID_hash", pskID)
-	infoHash := suite.KDF.LabeledExtract(nil, "info_hash", info)
+	suiteID := suite.hpkeSuiteID()
+	pskIDHash := suite.KDF.LabeledExtract(nil, suiteID, "pskID_hash", pskID)
+	infoHash := suite.KDF.LabeledExtract(nil, suiteID, "info_hash", info)
 
 	contextStruct := hpkeContext{suite.KEM.ID(), suite.KDF.ID(), suite.AEAD.ID(), mode, pskIDHash, infoHash}
```
Comment (rozbb) [inline hpke.go:179]: The first three fields of this struct have been killed

Comment (chris-wood) [reply hpke.go]: Great catch! Fixed.

**Comment by rozbb** [comment]
All tests passing (X25519 and P-256). This helped me catch a ton of errors in my own codebase.

File: hpke_test.go [hunk 8]
```diff
@@ -291,17 +302,20 @@ func (tv *testVector) UnmarshalJSON(data []byte) error {
...
 	modeRequiresSenderKey := (tv.mode == modeAuth || tv.mode == modeAuthPSK)
-	tv.skR = mustUnmarshalPriv(tv.t, tv.suite, raw.SKR, true)
-	tv.skS = mustUnmarshalPriv(tv.t, tv.suite, raw.SKS, modeRequiresSenderKey)
-	tv.skE = mustUnmarshalPriv(tv.t, tv.suite, raw.SKE, true)
+	tv.skR = mustDeserializePriv(tv.t, tv.suite, raw.SKR, true)
+	tv.skS = mustDeserializePriv(tv.t, tv.suite, raw.SKS, modeRequiresSenderKey)
+	tv.skE = mustDeserializePriv(tv.t, tv.suite, raw.SKE, true)
+
+	tv.pkR = mustDeserializePub(tv.t, tv.suite, raw.PKR, true)
+	tv.pkS = mustDeserializePub(tv.t, tv.suite, raw.PKI, modeRequiresSenderKey)
```
Comment (bifurcation) [inline hpke_test.go:310]: `s/PKI/PKS/`

File: hpke_test.go [hunk 9]
```diff
@@ -476,14 +490,47 @@ func verifyParameters(tv testVector, ctx cipherContext) {
 	assertBytesEqual(tv.t, tv.suite, "Incorrect parameter 'exporterSecret'", tv.exporterSecret, ctx.exporterSecret)
 }
 
+func verifyPublicKeysEqual(tv testVector, pkX, pkY KEMPublicKey) {
```
Comment (bifurcation) [inline hpke_test.go:493]: Better to define an `Equals()` method on `KEMPublicKey` and `KEMPrivateKey`?

File: hpke_test.go [hunk 10]
```diff
@@ -476,14 +490,47 @@ func verifyParameters(tv testVector, ctx cipherContext) {
...
+	pkXm := mustSerializePub(tv.suite, pkX)
+	pkYm := mustSerializePub(tv.suite, pkY)
+	assertBytesEqual(tv.t, tv.suite, "Incorrect public key", []byte(pkXm), []byte(pkYm))
```
Comment (bifurcation) [inline hpke_test.go:496]: Note for later: For other Go projects, I've been replacing custom assertions with github.com/stretchr/testify/require

File: hpke.go (hunk 7 above)
Comment (dmcardle) [reply hpke.go]: Passing!  I had made the same mistake, so I made another pass through https://github.com/cfrg/draft-irtf-cfrg-hpke/pull/127, but I didn't find any discrepancies.

File: hpke.go [hunk 11]
```diff
@@ -214,10 +215,10 @@ func newCipherContext(suite CipherSuite, setupParams setupParameters, contextPar
 		return cipherContext{}, err
 	}
 
-	return cipherContext{key, nonce, exporterSecrert, aead, 0, suite.KDF, nil, setupParams, contextParams}, nil
+	return cipherContext{key, nonce, exporterSecrert, aead, 0, suite, nil, setupParams, contextParams}, nil
```
Comment (dmcardle) [inline hpke.go:218]: This "secrert" typo has been bugging me :)

File: hpke.go [hunk 12]
```diff
@@ -66,6 +67,14 @@ type CipherSuite struct {
 	AEAD AEADScheme
 }
 
+func (suite CipherSuite) hpkeSuiteID() []byte {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke.go:70]: That works for me!

File: hpke.go [hunk 13]
```diff
@@ -90,42 +99,36 @@ func logVal(name string, value []byte) {
...
-		ok = gotPKIm && gotPSK
+	gotPSK := !bytes.Equal(psk, defaultPSK)
+	gotPSKID := !bytes.Equal(pskID, defaultPSKID)
+	if gotPSK != gotPSKID {
+		return fmt.Errorf("Inconsistent PSK inputs [%d] [%v] [%v]", mode, gotPSK, gotPSKID)
 	}
 
-	if !ok {
-		return fmt.Errorf("Invalid configuration [%d] [%v] [%v]", mode, gotPKIm, gotPSK)
+	if gotPSK {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke.go:120]: Nice -- I like that simplification. I'll take it.

File: hpke.go [hunk 14]
```diff
@@ -90,42 +99,36 @@ func logVal(name string, value []byte) {
...
+	return []byte{}
 }
 
 func defaultPSKID(suite CipherSuite) []byte {
 	return []byte{}
 }
 
-func verifyMode(suite CipherSuite, mode HPKEMode, psk, pskID, pkSm []byte) error {
-	defaultPKIm := defaultPKIm(suite)
+func verifyPSKInputs(suite CipherSuite, mode HPKEMode, psk, pskID []byte) error {
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply hpke.go:110]: Agreed. I'll drop "HPKE" from the type name.

File: hpke_test.go (hunk 8 above)
Comment (chris-wood) [reply hpke_test.go:310]: 👍

File: hpke.go (hunk 11 above)
Comment (chris-wood) [reply hpke.go:218]: Hah! Nice. I'll fix it.

File: crypto_test.go [hunk 15]
```diff
@@ -18,16 +18,19 @@ func randomBytes(size int) []byte {
 
 func TestKEMSchemes(t *testing.T) {
 	schemes := []KEMScheme{
-		&dhkemScheme{group: x25519Scheme{}, KDF: hkdfScheme{hash: crypto.SHA256}},
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply crypto_test.go:21]: Yeah, we can. I'll make that change.

File: hpke_test.go (hunk 9 above)
Comment (chris-wood) [reply hpke_test.go:493]: I opted to keep this as is for consistency with the other checks we do.

**Pull request merged**